File: android/src/main/java/org/reallife/capacitor/zxingscanner/ZxingScannerPlugin.java
package org.reallife.capacitor.zxingscanner;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;

import androidx.activity.result.ActivityResult;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.ActivityCallback;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;

@CapacitorPlugin(
        name = "CapacitorBarcodeScanner",
        permissions = {
                @Permission(strings = {Manifest.permission.CAMERA}, alias = "camera")
        }
)
public class ZxingScannerPlugin extends Plugin {

    @PluginMethod
    public void scanBarcode(PluginCall call) {
        if (getPermissionState("camera") != com.getcapacitor.PermissionState.GRANTED) {
            requestPermissionForAlias("camera", call, "cameraPermissionCallback");
        } else {
            startScannerActivity(call);
        }
    }

    @PermissionCallback
    private void cameraPermissionCallback(PluginCall call) {
        if (getPermissionState("camera") == com.getcapacitor.PermissionState.GRANTED) {
            startScannerActivity(call);
        } else {
            call.reject("Camera permission denied");
        }
    }

    private void startScannerActivity(PluginCall call) {
        Intent intent = new Intent(getContext(), ScannerActivity.class);
        startActivityForResult(call, intent, "scanBarcodeResult");
    }

    @ActivityCallback
    private void scanBarcodeResult(PluginCall call, ActivityResult result) {
        if (result.getResultCode() == Activity.RESULT_OK && result.getData() != null) {
            String text = result.getData().getStringExtra(ScannerActivity.EXTRA_SCAN_RESULT);
            if (text != null) {
                JSObject ret = new JSObject();
                ret.put("ScanResult", text);
                ret.put("format", "QR_CODE");
                call.resolve(ret);
            } else {
                call.reject("No scan result");
            }
        } else {
            call.reject("Scanner cancelled");
        }
    }
}

File: ios/Sources/CapacitorZxingScanner/ZxingScannerPlugin.swift
import Foundation
import Capacitor
import AVFoundation

@objc(CapacitorBarcodeScannerPlugin)
public class ZxingScannerPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CapacitorBarcodeScannerPlugin"
    public let jsName = "CapacitorBarcodeScanner"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "scanBarcode", returnType: CAPPluginReturnPromise)
    ]

    @objc func scanBarcode(_ call: CAPPluginCall) {
        AVCaptureDevice.requestAccess(for: .video) { [weak self] granted in
            guard let self = self else { return }
            guard granted else {
                call.reject("Camera permission denied")
                return
            }
            DispatchQueue.main.async {
                self.presentScanner(call)
            }
        }
    }

    private func presentScanner(_ call: CAPPluginCall) {
        let scanner = ScannerViewController()
        scanner.modalPresentationStyle = .fullScreen
        scanner.completion = { [weak self] result in
            guard let self = self else { return }
            if let text = result {
                call.resolve(["ScanResult": text, "format": "QR_CODE"])
            } else {
                call.reject("Scanner cancelled")
            }
        }
        bridge?.viewController?.present(scanner, animated: true)
    }
}
